Frameworks/Starboard/EbrStorageFile.cpp

```diff
@@ -0,0 +1,351 @@
+
+    return success ? 0 : -1;
+}
+
+int EbrStorageFile::Stat(struct stat* ret) {
+    std::lock_guard<std::recursive_mutex> lock(m_lock);
+
+    if (!ret) {
+        return -1;
+    }
+
+    struct stat toReturn{};
+
+    ComPtr<IStorageItem> storageItem;
+    if (FAILED(m_storageFile.As(&storageItem))) {
+        return -1;
+    }
+
+    ComPtr<IBasicProperties> basicProperties;
+    if (FAILED(AwaitResult(storageItem, &IStorageItem::GetBasicPropertiesAsync, &basicProperties))) {
+        return -1;
+    }
+
+    unsigned long long size{};
+    if (FAILED(basicProperties->get_Size(&size))) {
+        return -1;
+    }
+
+    // TODO 1099: Add more properites that we can fill in stat.
+    toReturn.st_size = (decltype(toReturn.st_size))size;
+
+    // Just assume its a regular file.
+    toReturn.st_mode = S_IFREG;
+
+    *ret = toReturn;
+
+    return 0;
+}
+
+int EbrStorageFile::Stat64i32(struct _stat64i32* ret) {
+    std::lock_guard<std::recursive_mutex> lock(m_lock);
+
+    if (!ret) {
+        return -1;
+    }
+
+    struct _stat64i32 toReturn{};
+
+    ComPtr<IStorageItem> storageItem;
+    if (FAILED(m_storageFile.As(&storageItem))) {
+        return -1;
+    }
+
+    ComPtr<IBasicProperties> basicProperties;
+    if (FAILED(AwaitResult(storageItem, &IStorageItem::GetBasicPropertiesAsync, &basicProperties))) {
+        return -1;
+    }
+
+    unsigned long long size{};
+    if (FAILED(basicProperties->get_Size(&size))) {
+        return -1;
+    }
+
+    // TODO 1099: Add more properites that we can fill in stat.
+    toReturn.st_size = (decltype(toReturn.st_size))size;
+
+    // Just assume its a regular file.
+    toReturn.st_mode = S_IFREG;
+
+    *ret = toReturn;
+
+    return 0;
+}
+
+intptr_t EbrStorageFile::GetOSFHandle() {
+    return -1;
+}
+
+int EbrStorageFile::Read(void* dest, size_t count) {
+    std::lock_guard<std::recursive_mutex> lock(m_lock);
+
+    if (!dest) {
+        return -1;
+    }
+
+    unsigned long long currentPosition{};
+    if (FAILED(m_randomAccessStream->get_Position(&currentPosition))) {
+        return -1;
+    }
+
+    ComPtr<IInputStream> inStream;
+    if (FAILED(m_randomAccessStream->GetInputStreamAt(currentPosition, &inStream))) {
+        return -1;
+    }
+
+    std::vector<unsigned char> buffer(count);
+    ComPtr<IBuffer> rawDataBuffer;
+    if (FAILED(BufferFromRawData(&rawDataBuffer, buffer.data(), buffer.size()))) {
+        return -1;
+    }
+
+    ComPtr<IBuffer> dataRead;
+    if (FAILED(AwaitProgressComplete(inStream, &IInputStream::ReadAsync, rawDataBuffer.Get(), buffer.size(), InputStreamOptions_None, &dataRead))) {
+        return -1;
+    }
+
+    unsigned int length = 0;
+    if (FAILED(dataRead->get_Length(&length))) {
+        return -1;
+    }
+
+    ComPtr<IBufferByteAccess> byteAccess;
+    if (FAILED(dataRead.As(&byteAccess))) {
+        return -1;
+    }
+
+    unsigned char* rawBytes = nullptr;
+    if (FAILED(byteAccess->Buffer(&rawBytes))) {
+        return -1;
+    }
```

Comment:
again, the macro introduction of RETURN_RESULT_IF_FAILED(.......,-1);
would help allot here.